From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 22 Jun 2024 20:54:50 +0800
Subject: [PATCH] Rebrand


diff --git a/build.gradle.kts b/build.gradle.kts
index 7aa138f7123dcfb7d1d9ae9efcf2359522f5d2fd..ab1116a3abbce6c3c5a6fb484a00530291fd0153 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -135,6 +135,9 @@ val generateApiVersioningFile by tasks.registering {
 tasks.withType<JavaCompile> {
     val compilerArgs = options.compilerArgs
     compilerArgs.add("--add-modules=jdk.incubator.vector")
+    compilerArgs.add("-Xlint:-deprecation") // Vine
+    options.isWarnings = false // Vine
+
 }
 // Pufferfish End
 
diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 7107eabf02d9d9b69d97c978ba421ad52031ec3f..5835ffe0ad932f3a6fa0617f4f897a95eb9c126f 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -22,7 +22,7 @@ public interface ServerBuildInfo {
     /**
      * The brand id for Purpur.
      */
-    Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
+    Key BRAND_PURPUR_ID = Key.key("tranic", "vine");
     // Purpur end
 
     /**
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index 0ec64c622ebfaeb9bc607ba774620ccc7233021b..852161d3549ffe9afc5bb6184d9af8259ee1d9bb 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -1,16 +1,7 @@
 package org.bukkit.command.defaults;
 
-import com.google.common.base.Charsets;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
-import com.google.common.io.Resources;
-import com.google.gson.Gson;
-import com.google.gson.JsonObject;
-import com.google.gson.JsonSyntaxException;
-import java.io.BufferedReader;
-import java.io.IOException;
-import java.net.URL;
-import java.net.URLEncoder;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashSet;
@@ -259,7 +250,7 @@ public class VersionCommand extends BukkitCommand {
         // Purpur start
         int distance = getVersionFetcher().distance();
         final Component message = Component.join(net.kyori.adventure.text.JoinConfiguration.separator(Component.newline()),
-            ChatColor.parseMM("<grey>Current ShreddedPaper Version: %s%s*", distance == 0 ? "<green>" : distance > 0 ? "<yellow>" : "<red>", Bukkit.getVersion()),
+            ChatColor.parseMM("<grey>Current Vine Server Version: %s%s*", distance == 0 ? "<green>" : distance > 0 ? "<yellow>" : "<red>", Bukkit.getVersion()),
             // Purpur end
             msg
         );
@@ -281,25 +272,4 @@ public class VersionCommand extends BukkitCommand {
             versionLock.unlock();
         }
     }
-
-    private static int getDistance(@NotNull String repo, @NotNull String hash) {
-        try {
-            BufferedReader reader = Resources.asCharSource(
-                    new URL("https://hub.spigotmc.org/stash/rest/api/1.0/projects/SPIGOT/repos/" + repo + "/commits?since=" + URLEncoder.encode(hash, "UTF-8") + "&withCounts=true"),
-                    Charsets.UTF_8
-            ).openBufferedStream();
-            try {
-                JsonObject obj = new Gson().fromJson(reader, JsonObject.class);
-                return obj.get("totalCount").getAsInt();
-            } catch (JsonSyntaxException ex) {
-                ex.printStackTrace();
-                return -1;
-            } finally {
-                reader.close();
-            }
-        } catch (IOException e) {
-            e.printStackTrace();
-            return -1;
-        }
-    }
 }
