From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Tue, 13 Aug 2024 10:11:40 +0800
Subject: [PATCH] Vine Config


diff --git a/src/main/java/one/tranic/vine/config/VineConfig.java b/src/main/java/one/tranic/vine/config/VineConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..80ba67d0107cbc46542ae5b2d62111c1c24c14e3
--- /dev/null
+++ b/src/main/java/one/tranic/vine/config/VineConfig.java
@@ -0,0 +1,78 @@
+package one.tranic.vine.config;
+
+public class VineConfig {
+    /**This is some fix*/
+    public static class Fix {
+        /**This is just a testing option, turning it on or off will not change anything.*/
+        public static boolean network = false;
+
+        /**Ensures correct destinations when travelling back and forth through Nether Portals in Multiplayer.*/
+        public static boolean netherPortal = false;
+
+        /**
+         * Fixed a bug in Paper where anvils could become drops at y=0 or y=-1.
+         * This may not be useful as the implementation is crude, you can install the fix plugin to fix it additionally if you can.
+         */
+        public static boolean anvilDrop = true;
+        public static boolean anvilDropLogger = false;
+
+        /**Fixed the bug that the display frame in water was destroyed by TNT.*/
+        public static boolean explosionBreaksItemFrameInWater = false;
+
+        /**Fixes some entities not bouncing on slime blocks and getting stuck*/
+        public static boolean incorrectBounceLogic = false;
+    }
+
+    public static class Feature {
+        public static boolean vanillaHopper = false;
+
+        public static boolean flattenTriangularDistribution = false;
+
+        /**zero tick plants is back now!*/
+        public static boolean zeroTickPlants = false;
+
+        /**
+         * Switch to a 1024-bit secure seed to protect it from being cracked.
+         * Destroy the ambitions of those cheaters.
+         * Of course I haven't confirmed whether it's really possible XD
+         * */
+        public static boolean secureSeed = false;
+
+        /**This just removes spark from being loaded by default, if you disable it, or install spark yourself, it will still be loaded.*/
+        public static boolean disableSpark = true;
+
+        /**
+         * This can prevent some devices that utilize vanilla abilities from being
+         * damaged when the server crashes or shuts down suddenly, but it is not a
+         * panacea, so take precautions.*/
+        public static class DSE {
+            /**Do not save tnt entities*/
+            public static boolean primedTNT = false;
+
+            /**Do not save falling blocks*/
+            public static boolean fallingBlock = false;
+        }
+    }
+
+    public static class Optimize {
+        /**As the name suggests, it can slightly improve world generation performance, maybe.*/
+        public static boolean noiseGeneration = true;
+
+        /**Optimize stacked minecarts*/
+        public static class Minecart {
+            public static boolean  enabled = true;
+            public static int skipTickCount = 30;
+        }
+
+        /**
+         * Entity re-pathing optimization
+         * If the entity has already determined the path, it will not search the path as frequently.
+         */
+        public static boolean mobRePathfinding = true;
+
+        public static class TurtleEggSearch {
+            public static boolean enabled = false;
+            public static int percentChance = 40;
+        }
+    }
+}
