From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Mon, 19 Aug 2024 00:17:10 +0800
Subject: [PATCH] Kotlin api


diff --git a/build.gradle.kts b/build.gradle.kts
index ab1116a3abbce6c3c5a6fb484a00530291fd0153..7cb844b49caa3b98f3fefcc3bb90ce51780adfa0 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -58,6 +58,16 @@ dependencies {
     implementation("org.ow2.asm:asm-commons:9.7")
     // Paper end
 
+    // Vine start
+    api("org.jetbrains.kotlin:kotlin-stdlib:2.0.21")
+    api("org.jetbrains.kotlin:kotlin-reflect:2.0.21")
+    api("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.9.0")
+    api("org.jetbrains.kotlinx:kotlinx-io-core:0.5.4")
+    api("org.jetbrains.kotlinx:atomicfu:0.26.0")
+    api("org.jetbrains.kotlinx:kotlinx-serialization-core:1.7.3")
+    api("org.jetbrains.kotlinx:kotlinx-serialization-json:1.7.3")
+    // Vine end
+
     api("org.apache.maven:maven-resolver-provider:3.9.6") // Paper - make API dependency for Paper Plugins
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.18")
     compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.18")
