From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Wed, 21 Aug 2024 21:26:00 +0800
Subject: [PATCH] Add some required dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index 5ecc93bb8a1828bbe44e0cc76c72eeb41b120001..0dfe4f14099d5e3ff4da943d32bd14a1e788d903 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -85,6 +85,27 @@ dependencies {
     implementation("org.jetbrains.kotlinx:kotlinx-serialization-json:1.7.2")
     // Vine end
 
+    // Vine start
+    implementation("org.xerial.snappy:snappy-java:1.1.10.6")
+    implementation("com.github.luben:zstd-jni:1.5.6-4")
+    implementation("org.lz4:lz4-java:1.8.0")
+
+    implementation("org.reflections:reflections:0.10.2")
+    implementation("org.simdjson:simdjson-java:0.3.0")
+
+    val brotli4jVersion = "1.16.0"
+    implementation ( group = "com.aayushatharva.brotli4j", name = "brotli4j", version = brotli4jVersion )
+    runtimeOnly ( group = "com.aayushatharva.brotli4j", name = "native-windows-x86_64", version = brotli4jVersion )
+    runtimeOnly ( group = "com.aayushatharva.brotli4j", name = "native-windows-aarch64", version = brotli4jVersion )
+    runtimeOnly ( group = "com.aayushatharva.brotli4j", name = "native-linux-x86_64", version = brotli4jVersion )
+    runtimeOnly ( group = "com.aayushatharva.brotli4j", name = "native-linux-aarch64", version = brotli4jVersion )
+    runtimeOnly ( group = "com.aayushatharva.brotli4j", name = "native-linux-armv7", version = brotli4jVersion )
+    runtimeOnly ( group = "com.aayushatharva.brotli4j", name = "native-linux-ppc64le", version = brotli4jVersion )
+    runtimeOnly ( group = "com.aayushatharva.brotli4j", name = "native-linux-riscv64", version = brotli4jVersion )
+    runtimeOnly ( group = "com.aayushatharva.brotli4j", name = "native-linux-s390x", version = brotli4jVersion )
+    runtimeOnly ( group = "com.aayushatharva.brotli4j", name = "native-osx-x86_64", version = brotli4jVersion )
+    runtimeOnly ( group = "com.aayushatharva.brotli4j", name = "native-osx-aarch64", version = brotli4jVersion )
+    // Vine end
 }
 
 paperweight {
