From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Wed, 28 Aug 2024 09:07:38 +0800
Subject: [PATCH] Add vine and shreddedpaper configuration files to spark


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index ccd25a26cfd3b4075279bb89657563fb7f2fbd21..f1eed0eef796d3ecced91d6bf770e1aead019650 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1175,6 +1175,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
             org.spigotmc.WatchdogThread.tick(); // Paper
             org.spigotmc.WatchdogThread.hasStarted = true; // Paper
+            one.tranic.vine.config.ConfigUtil.addToSpark(); // Vine
             Arrays.fill( this.recentTps, 20 );
             // Paper start - further improve server tick loop
             long tickSection = Util.getNanos();
diff --git a/src/main/kotlin/one/tranic/vine/config/ConfigUtil.kt b/src/main/kotlin/one/tranic/vine/config/ConfigUtil.kt
index 655dcd125a58aec84ee93a90cf087b0b10052830..a0e735da72644446a5fba7832d1e33ed62f8419b 100644
--- a/src/main/kotlin/one/tranic/vine/config/ConfigUtil.kt
+++ b/src/main/kotlin/one/tranic/vine/config/ConfigUtil.kt
@@ -55,6 +55,13 @@ object ConfigUtil {
         }
     }
 
+    @JvmStatic
+    fun addToSpark() {
+        if (!VineConfig.Feature.disableSpark && Bukkit.getServer().getPluginManager().getPlugin("spark") != null) {
+            System.setProperty("spark.serverconfigs.extra", "vine.yml,shreddedpaper.yml")
+        }
+    }
+
     private fun addDefault(configFile: File) {
         configuration.options().setHeader(
             listOf(
