From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Tue, 3 Sep 2024 22:59:12 +0800
Subject: [PATCH] Command Utils


diff --git a/src/main/kotlin/one/tranic/vine/commands/Command.kt b/src/main/kotlin/one/tranic/vine/commands/Command.kt
new file mode 100644
index 0000000000000000000000000000000000000000..9087265af9caf04b5bf8dee0d001d65a2163f4ed
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/commands/Command.kt
@@ -0,0 +1,5 @@
+package one.tranic.vine.commands
+
+object Command {
+    private val clazz: ArrayList<org.bukkit.command.Command> = ArrayList()
+}
\ No newline at end of file
