From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Wed, 4 Sep 2024 09:35:57 +0800
Subject: [PATCH] VineLogger


diff --git a/src/main/kotlin/one/tranic/vine/util/VineLogger.kt b/src/main/kotlin/one/tranic/vine/util/VineLogger.kt
new file mode 100644
index 0000000000000000000000000000000000000000..9c3a9a5c74d16b2cd650188c3d2a386d9e885da4
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/util/VineLogger.kt
@@ -0,0 +1,54 @@
+package one.tranic.vine.util
+
+import net.minecraft.server.MinecraftServer
+import org.slf4j.Logger
+
+object VineLogger {
+    private val logger: Logger = MinecraftServer.LOGGER
+
+    fun get(): Logger {
+        return logger
+    }
+
+    fun info(msg: String) {
+        logger.info(msg)
+    }
+
+    fun info(msg: String, vararg args: Any) {
+        logger.info(msg, args)
+    }
+
+    fun info(msg: kotlin.collections.List<String>) {
+        for (str in msg) {
+            logger.info(str)
+        }
+    }
+
+    fun warn(msg: String) {
+        logger.warn(msg)
+    }
+
+    fun warn(msg: String, vararg args: Any) {
+        logger.warn(msg, args)
+    }
+
+    fun warn(msg: kotlin.collections.List<String>) {
+        for (str in msg) {
+            logger.warn(str)
+        }
+    }
+
+    fun error(msg: String) {
+        logger.error(msg)
+    }
+
+    fun error(msg: String, vararg args: Any) {
+        logger.error(msg, args)
+    }
+
+    fun error(msg: kotlin.collections.List<String>) {
+        for (str in msg) {
+            logger.error(str)
+        }
+    }
+}
\ No newline at end of file
