From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Tue, 22 Oct 2024 11:16:12 +0800
Subject: [PATCH] List Utils


diff --git a/src/main/java/one/tranic/vine/util/List.java b/src/main/java/one/tranic/vine/util/List.java
new file mode 100644
index 0000000000000000000000000000000000000000..2f782f7bca06ded6000c697c0d937352e8b6c710
--- /dev/null
+++ b/src/main/java/one/tranic/vine/util/List.java
@@ -0,0 +1,28 @@
+package one.tranic.vine.util;
+
+import com.google.common.collect.Iterators;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class List {
+    public static <T extends @Nullable Object> java.util.List<T> newList() {
+        return new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+    }
+
+    public static <T extends @Nullable Object> java.util.List<T> newList(Iterable<? extends T> elements) {
+        checkNotNull(elements);
+        if (elements instanceof Collection) {
+            return newList((Collection<? extends T>) elements);
+        }
+        java.util.List<T> list = newList();
+        Iterators.addAll(list, elements.iterator());
+        return list;
+    }
+
+    public static <T extends @Nullable Object> java.util.List<T> newList(Collection<? extends T> c) {
+        return new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(c);
+    }
+}
