From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sun, 23 Jun 2024 21:01:06 +0800
Subject: [PATCH] Vine Config


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index a7ffbbfd2a9fdeae0953ec8753ce48e93ccce086..9e58bce64d46a1ac12bcc067a07883069a27e482 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -121,6 +121,8 @@ public class Main {
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
+            one.tranic.vine.config.ConfigUtil.load(); // Vine - load config
+
             ShreddedPaperConfigurationLoader.init((File) optionset.valueOf("shreddedpaper-settings")); // ShreddedPaper
 
             // Purpur start - load config files early
diff --git a/src/main/java/one/tranic/vine/config/package-info.java b/src/main/java/one/tranic/vine/config/package-info.java
new file mode 100644
index 0000000000000000000000000000000000000000..45f4c24cc689252513752e7b1cfa6736aa465f6b
--- /dev/null
+++ b/src/main/java/one/tranic/vine/config/package-info.java
@@ -0,0 +1,10 @@
+/**
+ * Vine Server configuration file package
+ * <p>
+ * If the public method needs to be called by Java, it needs to be annotated with @JvmStatic.
+ * If it is a field exposed to Java, it needs to be annotated with @JvmField.
+ * <p>
+ * If you need to modify this file, you'll need to modify the Vine Config commit using
+ * `git rebase -i HEAD~N` rather than patching it in a new patch.
+ * */
+package one.tranic.vine.config;
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/ConfigUtil.kt b/src/main/kotlin/one/tranic/vine/config/ConfigUtil.kt
new file mode 100644
index 0000000000000000000000000000000000000000..9d8671f2902cfce4ce1f69c808907500069d27ba
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/ConfigUtil.kt
@@ -0,0 +1,86 @@
+package one.tranic.vine.config
+
+import com.aayushatharva.brotli4j.Brotli4jLoader
+import one.tranic.vine.region.Compression
+import one.tranic.vine.region.Format
+import one.tranic.vine.util.VineLogger
+import org.bukkit.configuration.file.YamlConfiguration
+import java.io.File
+
+object ConfigUtil {
+    private lateinit var configuration: YamlConfiguration
+
+    fun getConfiguration() : YamlConfiguration {
+        return configuration
+    }
+
+    /**
+     * Load the configuration file.
+     *
+     * This method should be called twice on the first load and only once
+     * thereafter to ensure that annotations are generated correctly.
+     *
+     * To be honest, I really want to split it up, such as using annotations
+     * like @Comment and @Path, and using reflection to automatically register
+     * configuration items, which may be much more convenient. However,
+     * perhaps because my method is wrong, it does not take effect in Kotlin,
+     * so I still register these configuration items manually.
+     * */
+    @JvmStatic
+    fun reload() {
+        runCatching {
+            val configFile = File("vine.yml")
+            if (!configFile.exists()) {
+                configFile.createNewFile()
+            }
+            configuration = YamlConfiguration.loadConfiguration(configFile)
+            addDefault(configFile)
+            one.tranic.vine.config.module.Module.load()
+        }.getOrElse {
+            throw RuntimeException(it)
+        }
+    }
+
+    @JvmStatic
+    fun load() {
+        reload()
+        reload()
+        // If SECTORFILE is enabled and BROTLI is selected as the compression format,
+        // initialize the dependent libraries.
+        if (Format.equals(Format.SECTORFILE) && Compression.equals(Compression.BROTLI)) {
+            Brotli4jLoader.ensureAvailability()
+            VineLogger.warn(
+                listOf(
+                    "****************************",
+                    "Brotli in SectorFile currently faces poor performance issues,",
+                    "it is recommended to choose ZSTD instead of Brotli",
+                    "****************************"
+                )
+            )
+        }
+        VineLogger.warn(
+            listOf(
+                "******************************************",
+                "Since Vine is compatible with multiple region formats, it is not possible to automatically convert between them. ",
+                "Currently, it only supports conversion between ANVIL and SectorFile.",
+                "If you have relevant needs, please download the relevant tools to convert region formats.",
+                "******************************************"
+            )
+        )
+    }
+
+    private fun addDefault(configFile: File) {
+        configuration.options().setHeader(
+            listOf(
+                "Vine Abstract Configuration",
+                "Vine Github: https://github.com/LevelTranic/Vine",
+                "It is recommended to always keep the latest version, download it at https://tranic.one/downloads/vine"
+            )
+        )
+
+        one.tranic.vine.config.module.Module.default()
+
+        configuration.options().copyDefaults(true)
+        configuration.save(configFile)
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/impl/ConfigImpl.kt b/src/main/kotlin/one/tranic/vine/config/impl/ConfigImpl.kt
new file mode 100644
index 0000000000000000000000000000000000000000..173945a287f4d66ffc0373d0e0db9ba553bcce6e
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/impl/ConfigImpl.kt
@@ -0,0 +1,6 @@
+package one.tranic.vine.config.impl
+
+interface ConfigImpl {
+    fun default()
+    fun load()
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/Locker.kt b/src/main/kotlin/one/tranic/vine/config/module/Locker.kt
new file mode 100644
index 0000000000000000000000000000000000000000..d336332ccdd0fb8d78da01004981fea5716fa29e
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/Locker.kt
@@ -0,0 +1,12 @@
+package one.tranic.vine.config.module
+
+import one.tranic.vine.region.Compression
+import one.tranic.vine.region.Format
+
+object Locker {
+    var compressionFormat: Compression? = null
+    var regionFormat: Format? = null
+    var regionCompressionLevel: Int? = null
+    var secureSeed: Boolean? = null
+    var virtualThread: Boolean? = null
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/Module.kt b/src/main/kotlin/one/tranic/vine/config/module/Module.kt
new file mode 100644
index 0000000000000000000000000000000000000000..3ccd0ab0e830c91575702bc23d706b4f17e5f821
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/Module.kt
@@ -0,0 +1,34 @@
+package one.tranic.vine.config.module
+
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.util.Reflect
+
+object Module : ConfigImpl {
+    private val clazz: ArrayList<ConfigImpl> = ArrayList()
+
+    private fun ArrayList<ConfigImpl>.addImpl(packageName: String) {
+        this.addAll(Reflect.findObjectClass(packageName))
+    }
+
+    init {
+        clazz.addImpl("one.tranic.vine.config.module.feature")
+        clazz.addImpl("one.tranic.vine.config.module.fix")
+        clazz.addImpl("one.tranic.vine.config.module.optimize")
+        clazz.addImpl("one.tranic.vine.config.module.optimize.entitygoal")
+        clazz.addImpl("one.tranic.vine.config.module.optimize.entitygoal.player")
+        clazz.addImpl("one.tranic.vine.config.module.optimize.entitygoal.mob")
+        clazz.addImpl("one.tranic.vine.config.module.optimize.entitygoal.mob.zombievillage")
+    }
+
+    override fun default() {
+        for (c in clazz) {
+            c.default()
+        }
+    }
+
+    override fun load() {
+        for (c in clazz) {
+            c.load()
+        }
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/feature/DontSaveEntity.kt b/src/main/kotlin/one/tranic/vine/config/module/feature/DontSaveEntity.kt
new file mode 100644
index 0000000000000000000000000000000000000000..9d7053f0d68b82ac5830881ede2cb5f99470c6ae
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/feature/DontSaveEntity.kt
@@ -0,0 +1,24 @@
+package one.tranic.vine.config.module.feature
+
+import one.tranic.vine.config.VineConfig.Feature
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object DontSaveEntity : ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "feature.dont-save-entity"
+    override fun default() {
+        addDefault(
+            "$BASEPATH.primed-tnt", false, listOf(
+                "Disable save primed tnt on chunk unloads.",
+                "Useful for redstone server, can prevent machine be exploded by TNT,",
+                "when player disconnected caused by Internet issue."
+            )
+        )
+        addDefault("$BASEPATH.falling-block", false)
+    }
+
+    override fun load() {
+        Feature.DSE.primedTNT = getBoolean("$BASEPATH.primed-tnt")
+        Feature.DSE.fallingBlock = getBoolean("$BASEPATH.falling-block")
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/feature/PlayerProfile.kt b/src/main/kotlin/one/tranic/vine/config/module/feature/PlayerProfile.kt
new file mode 100644
index 0000000000000000000000000000000000000000..4c9654fd67f8b68a30af044b0886e146412cf34b
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/feature/PlayerProfile.kt
@@ -0,0 +1,25 @@
+package one.tranic.vine.config.module.feature
+
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.VineConfig.Feature
+import one.tranic.vine.config.util.ConfigHelper
+
+object PlayerProfile: ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "feature.cache-player-profile-result"
+    override fun default() {
+        addDefault(
+            "$BASEPATH.enabled",
+            true,
+            listOf(
+                "Cache the player profile result on they first join.",
+                "It's useful if Mojang's verification server is down."
+            )
+        )
+        addDefault("$BASEPATH.timeout", 1440, "The timeout of the cache. Unit: Minutes.")
+    }
+
+    override fun load() {
+        Feature.cachePlayerProfileResult.enabled = getBoolean("$BASEPATH.enabled")
+        Feature.cachePlayerProfileResult.timeout = getInt("$BASEPATH.timeout")
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/feature/Region.kt b/src/main/kotlin/one/tranic/vine/config/module/feature/Region.kt
new file mode 100644
index 0000000000000000000000000000000000000000..25e58745122872a473fe5614979f061e4f0dddf7
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/feature/Region.kt
@@ -0,0 +1,86 @@
+package one.tranic.vine.config.module.feature
+
+import one.tranic.vine.config.VineConfig.Feature
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.module.Locker
+import one.tranic.vine.config.util.ConfigHelper
+import one.tranic.vine.region.Compression
+import one.tranic.vine.region.Format
+import one.tranic.vine.util.VineLogger
+
+object Region : ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "feature.region"
+    override fun default() {
+        addDefault(
+            "$BASEPATH.format",
+            "ANVIL",
+            listOf(
+                "Region Format: ANVIL (Default), LINEAR, SECTORFILE",
+                "LINEAR: https://github.com/xymb-endcrystalme/LinearRegionFileFormatTools",
+                "SECTORFILE: https://github.com/PaperMC/SectorTool/blob/master/SPECIFICATION.MD",
+                "If you are not worried about disk capacity, please do not use Linear, it is still",
+                "experimental and may cause data corruption."
+            )
+        )
+        addDefault(
+            "$BASEPATH.compression-format", "ZSTD", listOf(
+                "Only applicable to SECTORFILE, support list:",
+                "\"GZIP, ZLIB, NONE, LZ4, BROTLI, SNAPPY, ZSTD\"",
+                "Brotli is only available on TranicServer for Vine.",
+                "There is currently no tool to convert it back to ANVIL. Enable with caution!"
+            )
+        )
+        addDefault(
+            "$BASEPATH.compression-level",
+            1,
+            listOf(
+                "Only applicable to Linear, this is the compression level provided by ZSTD.",
+                "The available range is 1-22. When the compression level exceeds 14, ",
+                "the CPU pressure, chunk refresh time and memory usage will increase significantly, ",
+                "and high compression levels have the risk of losing chunk data, ",
+                "so we will limit it to the range of 1-15 to avoid some simple problems."
+            )
+        )
+        addDefault(
+            "$BASEPATH.flush-frequency",
+            5,
+            "Only applicable to Linear, chunk refresh interval, default is 5 (seconds)"
+        )
+    }
+
+    override fun load() {
+        getString("$BASEPATH.format") {
+            val b = Format.getFormat(it)
+            if (Locker.regionFormat == null) {
+                Locker.regionFormat = b
+                Format.setNow(b)
+            } else if (!b.equals(Locker.regionFormat)) {
+                VineLogger.error("Forbidden to modify the parameter \"$BASEPATH.format\" during runtime!!!")
+            }
+        }
+        getString("$BASEPATH.compression-format") {
+            val b = Compression.getCompression(it)
+            if (Locker.compressionFormat == null) {
+                Locker.compressionFormat = b
+                Compression.setNow(b)
+            } else if (!b.equals(Locker.compressionFormat)) {
+                VineLogger.error("Forbidden to modify the parameter \"$BASEPATH.compression-format\" during runtime!!!")
+            }
+        }
+        getInt("$BASEPATH.compression-level") {
+            if (Locker.regionCompressionLevel == null) {
+                Locker.regionCompressionLevel = if (it < 1) {
+                    1
+                } else if (it > 15) {
+                    15
+                } else {
+                    it
+                }
+                Feature.Region.compressionLevel = Locker.regionCompressionLevel!!
+            } else if (Locker.regionCompressionLevel != it) {
+                VineLogger.error("Forbidden to modify the parameter \"$BASEPATH.compression-level\" during runtime!!!")
+            }
+        }
+        Feature.Region.flushFrequency = getInt("$BASEPATH.flush-frequency")
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/feature/Value.kt b/src/main/kotlin/one/tranic/vine/config/module/feature/Value.kt
new file mode 100644
index 0000000000000000000000000000000000000000..fda5396e1e05a550859c80c0753678d61fc0eec5
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/feature/Value.kt
@@ -0,0 +1,89 @@
+package one.tranic.vine.config.module.feature
+
+import one.tranic.vine.config.VineConfig.Feature
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.module.Locker
+import one.tranic.vine.config.util.ConfigHelper
+import one.tranic.vine.util.VineLogger
+
+object Value : ConfigImpl, ConfigHelper() {
+    override fun default() {
+        addDefault("feature.disable-spark", true)
+        addDefault(
+            "feature.disable-respawn-ceiling",
+            false,
+            listOf(
+                "If the player is blocked on the top of the player's head, he will not try to rebirth ",
+                "the player on the ceiling.",
+                "This is suitable for those servers with ceilings."
+            )
+        )
+        addDefault(
+            "feature.flatten-triangular-distribution",
+            false,
+            listOf(
+                "Change all triangle distribution in Minecraft randomizers into uniform distribution",
+                "With that edge cases are more likely to happen"
+            )
+        )
+        addDefault("feature.liquid-physics", false, "The new liquid physics system may affect some mechanisms.")
+        addDefault(
+            "feature.secure-seed", false, listOf(
+                "Changes the seed from 64bit to 1024bit, this will strictly secure your seed,",
+                "making it almost impossible to crack.",
+                "Warning: You must delete existing maps to switch the state of the secure seed,",
+                "and do not modify it during hot loading, otherwise the server will crash and",
+                "cannot continue to load maps."
+            )
+        )
+        addDefault(
+            "feature.try-virtual-threads",
+            false,
+            listOf(
+                "Try virtual threads. Once enabled, all explicit platform threads will ",
+                "be replaced with virtual threads as much as possible, which may have ",
+                "some adverse consequences. Please enable it with caution."
+            )
+        )
+        addDefault("feature.vanilla-hopper", false)
+        addDefault("feature.zero-tick-plants", false)
+    }
+
+    override fun load() {
+        Feature.disableSpark = getBoolean("feature.disable-spark")
+        Feature.disableRespawnCeiling = getBoolean("feature.disable-respawn-ceiling")
+        Feature.flattenTriangularDistribution = getBoolean("feature.flatten-triangular-distribution")
+        Feature.liquidPhysics = getBoolean("feature.liquid-physics")
+        getBoolean("feature.secure-seed") {
+            if (Locker.secureSeed == null) {
+                Feature.secureSeed = it
+                Locker.secureSeed = it
+                if (it) VineLogger.warn("** SecureSeed has been enabled, and this value cannot be modified before deleting the map.")
+            } else if (Locker.secureSeed != it) {
+                VineLogger.get().error("Forbidden to modify the parameter \"feature.secure-seed\" during runtime!!!")
+            }
+        }
+        getBoolean("feature.try-virtual-threads") {
+            if (Locker.virtualThread == null) {
+                if (it) {
+                    VineLogger.warn(
+                        listOf(
+                            "********",
+                            "Virtual threads enabled.",
+                            "Will try to offload threads other than ShreddedPaper and ChunkThread to dummy threads",
+                            "This feature is very experimental and may have some undesirable consequences",
+                            "********"
+                        )
+                    )
+                }
+                Feature.tryVirtualThreads = it
+                Locker.virtualThread = it
+            } else if (it != Locker.virtualThread) {
+                VineLogger.get()
+                    .error("Forbidden to modify the parameter \"feature.try-virtual-threads\" during runtime!!!")
+            }
+        }
+        Feature.vanillaHopper = getBoolean("feature.vanilla-hopper")
+        Feature.zeroTickPlants = getBoolean("feature.zero-tick-plants")
+    }
+}
diff --git a/src/main/kotlin/one/tranic/vine/config/module/fix/AnvilDrop.kt b/src/main/kotlin/one/tranic/vine/config/module/fix/AnvilDrop.kt
new file mode 100644
index 0000000000000000000000000000000000000000..927f3cfa8916dbe91c18bf619826588bf8e57911
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/fix/AnvilDrop.kt
@@ -0,0 +1,30 @@
+package one.tranic.vine.config.module.fix
+
+import one.tranic.vine.config.VineConfig.Fix
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object AnvilDrop : ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "fix.anvil-drop"
+    override fun default() {
+        addDefault(
+            "$BASEPATH.enabled",
+            true,
+            listOf(
+                "This fix will attempt to prevent the bug in https://github.com/PaperMC/Folia/issues/217, ",
+                "it may cause some unexpected effects but should be harmless."
+            )
+        )
+        addDefault(
+            "$BASEPATH.debug", false,
+            "Whether to print detailed information to the console when this issue is triggered"
+        )
+    }
+
+    override fun load() {
+        Fix.AnvilDrop.enabled = getBoolean("$BASEPATH.enabled")
+        Fix.AnvilDrop.debug = getBoolean("$BASEPATH.debug")
+
+    }
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/fix/Value.kt b/src/main/kotlin/one/tranic/vine/config/module/fix/Value.kt
new file mode 100644
index 0000000000000000000000000000000000000000..10b3cffdfc28bc77a3e3b4deace76743f7eb62a5
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/fix/Value.kt
@@ -0,0 +1,43 @@
+package one.tranic.vine.config.module.fix
+
+import one.tranic.vine.config.VineConfig.Fix
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object Value : ConfigImpl, ConfigHelper() {
+    override fun default() {
+        addDefault(
+            "fix.explosion-breaks-item-frame-in-water",
+            false,
+            listOf(
+                "Fixes Explosions being able to destroy item frames in water",
+                "Fixes: https://bugs.mojang.com/browse/MC-3697"
+            )
+        )
+        addDefault(
+            "fix.incorrect-bounce-logic",
+            false,
+            listOf(
+                "Fixes some entities not bouncing on slime blocks and getting stuck",
+                "Fixes: https://bugs.mojang.com/browse/MC-216985"
+            )
+        )
+        addDefault(
+            "fix.nether-portal",
+            false,
+            "Ensures correct destinations when travelling back and forth through Nether Portals in Multiplayer."
+        )
+        addDefault(
+            "fix.villagers-dont-release-memory",
+            false,
+        )
+    }
+
+    override fun load() {
+        Fix.explosionBreaksItemFrameInWater = getBoolean("fix.explosion-breaks-item-frame-in-water")
+        Fix.incorrectBounceLogic = getBoolean("fix.incorrect-bounce-logic")
+        Fix.netherPortal = getBoolean("fix.nether-portal")
+        Fix.villagersDontReleaseMemory = getBoolean("fix.villagers-dont-release-memory")
+    }
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/optimize/LoadChunks.kt b/src/main/kotlin/one/tranic/vine/config/module/optimize/LoadChunks.kt
new file mode 100644
index 0000000000000000000000000000000000000000..7e1e69647dd21abb5ed05cb5d06d8ccfbf4625ef
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/optimize/LoadChunks.kt
@@ -0,0 +1,20 @@
+package one.tranic.vine.config.module.optimize
+
+import one.tranic.vine.config.VineConfig.Optimize.loadChunks
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object LoadChunks: ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "optimize.load-chunks"
+    override fun default() {
+        addDefault(
+            "$BASEPATH.spawn-phantoms", true
+        )
+        addDefault("$BASEPATH.activate-climbing-entities", true)
+    }
+
+    override fun load() {
+        loadChunks.spawnPhantoms = getBoolean("$BASEPATH.spawn-phantoms")
+        loadChunks.activateClimbingEntities = getBoolean("$BASEPATH.activate-climbing-entities")
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/optimize/Minecart.kt b/src/main/kotlin/one/tranic/vine/config/module/optimize/Minecart.kt
new file mode 100644
index 0000000000000000000000000000000000000000..b4080b6ae0000746dbc5923140eecb8d568da05d
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/optimize/Minecart.kt
@@ -0,0 +1,25 @@
+package one.tranic.vine.config.module.optimize
+
+import one.tranic.vine.config.VineConfig.Optimize.Minecart
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object Minecart : ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "optimize.minecart"
+    override fun default() {
+        addDefault(
+            "$BASEPATH.enabled", true,
+            listOf(
+                "Enable this feature to handle large amount of stacked Minecart better.",
+                "By skipping tick collisions to reduce expense getting entities list",
+                "and bukkit event calls, useful for the anarchy server."
+            )
+        )
+        addDefault("$BASEPATH.skipTickCount", 30)
+    }
+
+    override fun load() {
+        Minecart.enabled = getBoolean("$BASEPATH.enabled")
+        Minecart.skipTickCount = getInt("$BASEPATH.skipTickCount")
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/optimize/TurtleEggSearch.kt b/src/main/kotlin/one/tranic/vine/config/module/optimize/TurtleEggSearch.kt
new file mode 100644
index 0000000000000000000000000000000000000000..fd309464e0a2eb6e7beaf42b11950ab987f04c64
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/optimize/TurtleEggSearch.kt
@@ -0,0 +1,29 @@
+package one.tranic.vine.config.module.optimize
+
+import one.tranic.vine.config.VineConfig.Optimize.TurtleEggSearch
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object TurtleEggSearch : ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "optimize.zombie-search-and-destroy-turtle-egg"
+    override fun default() {
+        addDefault(
+            "$BASEPATH.enabled",
+            false,
+            "If true, optimizes the expensive searching done to find a turtle egg to crush, for zombies/husks, by default fully disabling it"
+        )
+        addDefault(
+            "$BASEPATH.percent-chance",
+            40,
+            "Percent chance to run for task: zombie seeking turtle eggs to destroy"
+        )
+    }
+
+    override fun load() {
+        TurtleEggSearch.enabled =
+            getBoolean("$BASEPATH.enabled")
+        TurtleEggSearch.percentChance =
+            getInt("$BASEPATH.percent-chance")
+    }
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/optimize/Value.kt b/src/main/kotlin/one/tranic/vine/config/module/optimize/Value.kt
new file mode 100644
index 0000000000000000000000000000000000000000..166b6ff925f8db5e67a42edb917f081e84121b74
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/optimize/Value.kt
@@ -0,0 +1,41 @@
+package one.tranic.vine.config.module.optimize
+
+import one.tranic.vine.config.VineConfig.Optimize
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object Value : ConfigImpl, ConfigHelper() {
+    override fun default() {
+        addDefault("optimize.acquire-poi-for-stuck-entity", 60, "Reduce acquire POI for stuck entities")
+        addDefault(
+            "optimize.faster-structure-gen-future-sequencing",
+            false,
+            "May cause the inconsistent order of future compose tasks."
+        )
+        addDefault("optimize.floating-point-positive", false)
+        addDefault(
+            "optimize.mob-re-pathfinding",
+            true,
+            "If true, disables the frequent pathfinds done while already following a path"
+        )
+        addDefault("optimize.noise-generation", true)
+        addDefault(
+            "optimize.skip-self-raid-check",
+            false,
+            listOf(
+                "Whether to skip some raid checks.",
+                "Enabling this will significantly reduce the server's performance."
+            )
+        )
+    }
+
+    override fun load() {
+        Optimize.acquirePoiForStuckEntity = getInt("optimize.acquire-poi-for-stuck-entity")
+        Optimize.fasterStructureGenFutureSequencing = getBoolean("optimize.faster-structure-gen-future-sequencing")
+        Optimize.floatingPointPositive = getBoolean("optimize.floating-point-positive")
+        Optimize.mobRePathfinding = getBoolean("optimize.mob-re-pathfinding")
+        Optimize.noiseGeneration = getBoolean("optimize.noise-generation")
+        Optimize.skipSelfRaidCheck = getBoolean("optimize.skip-self-raid-check")
+    }
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/Value.kt b/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/Value.kt
new file mode 100644
index 0000000000000000000000000000000000000000..429be85566b1b9d1fb4d8ed43fecb55119bb7797
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/Value.kt
@@ -0,0 +1,16 @@
+package one.tranic.vine.config.module.optimize.entitygoal
+
+import one.tranic.vine.config.VineConfig.Optimize.entityGoal
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object Value : ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "optimize.entity-goal"
+    override fun default() {
+        addDefault("$BASEPATH.max-entities-to-reduce-pathfinding-rates", 2)
+    }
+
+    override fun load() {
+        entityGoal.maxEntitiesToReducePathfindingRates = getInt("$BASEPATH.max-entities-to-reduce-pathfinding-rates")
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/mob/Value.kt b/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/mob/Value.kt
new file mode 100644
index 0000000000000000000000000000000000000000..46915d92e061542c3820f0bf0bcec2d638024373
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/mob/Value.kt
@@ -0,0 +1,16 @@
+package one.tranic.vine.config.module.optimize.entitygoal.mob
+
+import one.tranic.vine.config.VineConfig.Optimize.entityGoal.mob
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object Value: ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "optimize.entity-goal.mob"
+    override fun default() {
+        addDefault("$BASEPATH.enemy-targeting-reduced-rate-percent-chance", 10)
+    }
+
+    override fun load() {
+        mob.enemyTargetingReducedRatePercentChance = getInt("$BASEPATH.enemy-targeting-reduced-rate-percent-chance")
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/mob/Wandering.kt b/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/mob/Wandering.kt
new file mode 100644
index 0000000000000000000000000000000000000000..18f0f672f1c312c4c702210556f68c6efa0e6060
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/mob/Wandering.kt
@@ -0,0 +1,22 @@
+package one.tranic.vine.config.module.optimize.entitygoal.mob
+
+import one.tranic.vine.config.VineConfig.Optimize.entityGoal.mob.wandering
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object Wandering : ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "optimize.entity-goal.mob.wandering"
+    override fun default() {
+        addDefault("$BASEPATH.delay", 100)
+        addDefault("$BASEPATH.optimize", true)
+        addDefault("$BASEPATH.percent-chance", 100)
+        addDefault("$BASEPATH.reduce-rate-multiplier", 6)
+    }
+
+    override fun load() {
+        wandering.delay = getInt("$BASEPATH.delay")
+        wandering.optimize = getBoolean("$BASEPATH.optimize")
+        wandering.percentChance = getInt("$BASEPATH.percent-chance")
+        wandering.reducedRateMultiplier = getInt("$BASEPATH.reduce-rate-multiplier")
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/mob/zombievillage/Raid.kt b/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/mob/zombievillage/Raid.kt
new file mode 100644
index 0000000000000000000000000000000000000000..abdd72014480f0fabce21060b2e5c47096225926
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/mob/zombievillage/Raid.kt
@@ -0,0 +1,18 @@
+package one.tranic.vine.config.module.optimize.entitygoal.mob.zombievillage
+
+import one.tranic.vine.config.VineConfig.Optimize.entityGoal.mob.zombieVillage.raid
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object Raid : ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "optimize.entity-goal.mob.zombie-village.raid"
+    override fun default() {
+        addDefault("$BASEPATH.optimize", true)
+        addDefault("$BASEPATH.chance", 33)
+    }
+
+    override fun load() {
+        raid.optimize = getBoolean("$BASEPATH.optimize")
+        raid.chance = getInt("$BASEPATH.chance")
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/player/ProximityRateReduce.kt b/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/player/ProximityRateReduce.kt
new file mode 100644
index 0000000000000000000000000000000000000000..2353ed83365f37fd1c74ce550c9c16cac8ac4ff5
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/module/optimize/entitygoal/player/ProximityRateReduce.kt
@@ -0,0 +1,20 @@
+package one.tranic.vine.config.module.optimize.entitygoal.player
+
+import one.tranic.vine.config.VineConfig.Optimize.entityGoal.player.proximityRateReduce
+import one.tranic.vine.config.impl.ConfigImpl
+import one.tranic.vine.config.util.ConfigHelper
+
+object ProximityRateReduce: ConfigImpl, ConfigHelper() {
+    private const val BASEPATH = "optimize.entity-goal.player.proximity-rate-reduce"
+    override fun default() {
+        addDefault("$BASEPATH.enabled", true)
+        addDefault("$BASEPATH.scan-rate", 40)
+        addDefault("$BASEPATH.cutoff", 12)
+    }
+
+    override fun load() {
+        proximityRateReduce.enabled = getBoolean("$BASEPATH.enabled")
+        proximityRateReduce.scanRate = getInt("$BASEPATH.scan-rate")
+        proximityRateReduce.cutoff = getInt("$BASEPATH.cutoff")
+    }
+}
\ No newline at end of file
diff --git a/src/main/kotlin/one/tranic/vine/config/util/ConfigHelper.kt b/src/main/kotlin/one/tranic/vine/config/util/ConfigHelper.kt
new file mode 100644
index 0000000000000000000000000000000000000000..ba49e9075843cacf74b43da8da84dbbe0fc725c7
--- /dev/null
+++ b/src/main/kotlin/one/tranic/vine/config/util/ConfigHelper.kt
@@ -0,0 +1,77 @@
+package one.tranic.vine.config.util
+
+import one.tranic.vine.config.ConfigUtil
+
+open class ConfigHelper {
+    // Vine start - Configuration file tools
+    fun addDefault(path: String, value: Any) {
+        ConfigUtil.getConfiguration().addDefault(path, value)
+    }
+
+    fun addDefault(path: String, value: Any, comment: String) {
+        ConfigUtil.getConfiguration().addDefault(path, value)
+        ConfigUtil.getConfiguration().setComments(path, listOf(comment))
+    }
+
+    fun addComment(path: String, comment: String) {
+        ConfigUtil.getConfiguration().setComments(path, listOf(comment))
+    }
+
+    fun addComment(path: String, comment: List<String>) {
+        ConfigUtil.getConfiguration().setComments(path, comment)
+    }
+
+    fun addDefault(path: String, value: Any, comment: List<String>) {
+        ConfigUtil.getConfiguration().addDefault(path, value)
+        ConfigUtil.getConfiguration().setComments(path, comment)
+    }
+
+    fun getBoolean(path: String): Boolean {
+        return ConfigUtil.getConfiguration().getBoolean(path)
+    }
+
+    fun getBoolean(path: String, function: (it: Boolean) -> Unit) {
+        function(getBoolean(path))
+    }
+
+    fun getDouble(path: String): Double {
+        return ConfigUtil.getConfiguration().getDouble(path)
+    }
+
+    fun getDouble(path: String, function: (it: Double) -> Unit) {
+        function(getDouble(path))
+    }
+
+    fun getInt(path: String): Int {
+        return ConfigUtil.getConfiguration().getInt(path)
+    }
+
+    fun getInt(path: String, function: (it: Int) -> Unit) {
+        function(getInt(path))
+    }
+
+    fun getString(path: String): String? {
+        return ConfigUtil.getConfiguration().getString(path)
+    }
+
+    fun getString(path: String, function: (it: String?) -> Unit) {
+        function(getString(path))
+    }
+
+    fun getList(path: String): MutableList<*>? {
+        return ConfigUtil.getConfiguration().getList(path)
+    }
+
+    fun getStringList(path: String): MutableList<String> {
+        return ConfigUtil.getConfiguration().getStringList(path)
+    }
+
+    fun getStringList(path: String, function: (it: MutableList<String>) -> Unit) {
+        function(getStringList(path))
+    }
+
+    fun getMapList(path: String): MutableList<MutableMap<*, *>> {
+        return ConfigUtil.getConfiguration().getMapList(path)
+    }
+    // Vine end - Configuration file tools
+}
\ No newline at end of file
