From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Mon, 12 Aug 2024 08:49:13 +0800
Subject: [PATCH] Divine: Fix MC-167242

See https://bugs.mojang.com/browse/MC-167242

diff --git a/src/main/java/net/minecraft/world/entity/npc/Villager.java b/src/main/java/net/minecraft/world/entity/npc/Villager.java
index 971f2d594ead143f34911eea857203aac5bdd0b5..d4579e9053a8ec153685ea90d0db0f8a33883fe5 100644
--- a/src/main/java/net/minecraft/world/entity/npc/Villager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/Villager.java
@@ -995,6 +995,11 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
                 world.addFreshEntityWithPassengers(entitywitch, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
                 // CraftBukkit end
                 this.releaseAllPois();
+
+                if (this.isSleeping()) {
+                    this.stopSleeping();
+                }
+
                 this.discard(EntityRemoveEvent.Cause.TRANSFORMATION); // CraftBukkit - add Bukkit remove cause
             } else {
                 super.thunderHit(world, lightning);
