From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Wed, 4 Sep 2024 16:35:46 +0800
Subject: [PATCH] Paper PR Update Velocity natives

Source in https://github.com/PaperMC/Paper/pull/11347

diff --git a/build.gradle.kts b/build.gradle.kts
index 4777451dc9378f1bed1c5f663cf444dabe42ff4e..0b808e2d4fa6bd403799cc07ba2fc952fef94c2e 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -41,7 +41,7 @@ dependencies {
     runtimeOnly("com.mysql:mysql-connector-j:8.3.0")
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
     // Paper start - Use Velocity cipher
-    implementation("com.velocitypowered:velocity-native:3.1.2-SNAPSHOT") {
+    implementation("com.velocitypowered:velocity-native:3.3.0-SNAPSHOT") {
         isTransitive = false
     }
     // Paper end - Use Velocity cipher
