From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cryptite <cryptite@gmail.com>
Date: Tue, 25 Apr 2023 08:21:03 -0500
Subject: [PATCH] Slice: noEntityCollisions for Entity


diff --git a/src/main/java/io/papermc/paper/util/CollisionUtil.java b/src/main/java/io/papermc/paper/util/CollisionUtil.java
index ee8e9c0e3690e78f3cc621ddfca89ea4256d4803..0466b04f24830539f2c927481c1c7995ae2f4cb1 100644
--- a/src/main/java/io/papermc/paper/util/CollisionUtil.java
+++ b/src/main/java/io/papermc/paper/util/CollisionUtil.java
@@ -1753,7 +1753,7 @@ public final class CollisionUtil {
     public static boolean getEntityHardCollisions(final CollisionGetter getter, final Entity entity, AABB aabb,
                                                   final List<AABB> into, final int collisionFlags, final Predicate<Entity> predicate) {
         final boolean checkOnly = (collisionFlags & COLLISION_FLAG_CHECK_ONLY) != 0;
-        if (!(getter instanceof EntityGetter entityGetter)) {
+        if ((entity != null && entity.noEntityCollisions) || !(getter instanceof EntityGetter entityGetter)) {
             return false;
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 16027780d6ce37cf2b069cc3a0af87972175a434..f83c55a65ce97f9944049d7e7a9cf763ea957054 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -548,6 +548,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
      * Overriding this field will cause memory leaks.
      */
     private final boolean hardCollides;
+    public boolean noEntityCollisions; // Slice
 
     private static final java.util.Map<Class<? extends Entity>, Boolean> cachedOverrides = java.util.Collections.synchronizedMap(new java.util.WeakHashMap<>());
     {
