From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Tue, 22 Oct 2024 11:49:35 +0800
Subject: [PATCH] Kiterino: Super dumb liquid ticks limiter


diff --git a/src/main/java/net/minecraft/world/level/material/FlowingFluid.java b/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
index 242c529c30e03484ce8da6b1929e3428c2b6772d..e715d8c87fa53c6de67583f866ca7d3cfe5abada 100644
--- a/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
+++ b/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
@@ -837,7 +837,14 @@ public abstract class FlowingFluid extends Fluid {
                 // CraftBukkit end
                 world.scheduleTick(pos, fluid1.getType(), i);
                 world.updateNeighborsAt(pos, iblockdata.getBlock());
+            } // Kiterino start - Super dumb liquid physics limiter to reduce the load
+            else if (one.tranic.vine.config.VineConfig.Feature.liquidPhysics.base) {
+                if (org.bukkit.Bukkit.isLagging() || org.bukkit.Bukkit.getCurrentTick() % 20 > 12) {
+                    world.scheduleTick(pos, this, 9);
+                    return;
+                }
             }
+            // Kiterino end
         }
 
         this.spread(world, pos, state);
